core/dnsserver/server.go

```diff
@@ -277,6 +285,11 @@ func (s *Server) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg)
 
 	// Wildcard match, if we have found nothing try the root zone as a last resort.
 	if h, ok := s.zones["."]; ok && h.pluginChain != nil {
+
+		// See comment above.
+		ctx = context.WithValue(ctx, plugin.ZoneCtx, ".")
+		ctx = context.WithValue(ctx, plugin.AddrCtx, s.Addr)
```

Comment:
What about making a type with both field and using a single context value? It doesn’t seem like the zone and address are ever accessed individually.